File: two_sum.py
# first bruteforce solution
def find_two_sum(arr, target):
    for i in range(len(arr)):
        for j in range(len(arr)):
            if arr[i] + arr[j] == target:
                return (i, j)
    return None

# final leetcode solution
def find_two_sum(nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        nums_map = {}

        for i in range(len(nums)):
            num_to_find = target - nums[i]
            if num_to_find in nums_map:
                return [nums_map[num_to_find], i]
            nums_map[nums[i]] = i

if __name__ == "__main__":
    #print(find_two_sum([1, 3, 7, 9, 2], 11)) # Solution exists
    #print(find_two_sum([1, 3, 7, 9, 2], 15)) # Solution does not exist
    #print(find_two_sum([], 15)) # Empty list
    #print(find_two_sum([1], 15)) # Single element list

    print(find_two_sum([2, 7, 11, 15], 9))

File: two_sum.cs
// bruteforce solution 199ms
public int[] TwoSum(int[] nums, int target) {
	for(int num1=0; num1<nums.Length; num1++)
	{
		for(int num2=0; num2<nums.Length; num2++)
		{
			if(num2 == num1)
			{
				continue;
			}
			if(nums[num1] + nums[num2] == target)
			{
				return new int[] {num1, num2};
			}
		}
	}
	return new int[0];
}

// better solution 115ms
public class Solution {
    public int[] TwoSum(int[] nums, int target) {
        Dictionary<int, int> numsMap = new Dictionary<int, int>();
        for(int i=0; i<nums.Length; i++)
        {
            if(numsMap.TryGetValue(target-nums[i], out int value))
            {
                return new int[] {i, value};
            }
            numsMap[nums[i]] = i;
        }
        return new int[0];
    }
}
